include/unifex/async_scope.hpp

```diff
@@ -0,0 +1,242 @@
+struct _receiver<Sender, Scheduler>::type final
+    : _receiver_base, has_scheduler<Scheduler> {
+  explicit type(
+      const Scheduler& scheduler,
+      inplace_stop_token stoken,
+      void* op,
+      async_scope* scope) noexcept
+      : _receiver_base{stoken, op, scope},
+        has_scheduler<Scheduler>{scheduler} {}
+
+  // receivers uniquely own themselves; we don't need any special move-
+  // construction behaviour, but we do need to ensure no copies are made
+  type(type&&) noexcept = default;
+
+  ~type() = default;
+
+  // it's just simpler to skip this
+  type& operator=(type&&) = delete;
+
+  void set_value() noexcept {
+    set_done();
+  }
+
+  void set_done() noexcept {
+    // we're about to delete this, so save the scope for later
+    auto scope = scope_;
+
+    using op_t = manual_lifetime<connect_result_t<Sender, type>>;
+
+    auto op = static_cast<op_t*>(op_);
+    op->destruct();
+    delete op;
+
+    record_done(scope);
+  }
+};
+
+struct async_scope {
+  async_scope() noexcept = default;
+
+  ~async_scope() {
+    [[maybe_unused]] auto state = opState_.load(std::memory_order_relaxed);
+
+    assert(is_stopping(state));
+    assert(op_count(state) == 0);
+  }
+
+  template (typename Sender, typename Scheduler)
+    (requires scheduler<Scheduler> AND
+     sender_to<Sender, receiver<Sender, remove_cvref_t<Scheduler>>>)
+  void submit(Sender&& sender, Scheduler&& scheduler) {
+    using receiver_t = receiver<Sender, remove_cvref_t<Scheduler>>;
+    using op_t = connect_result_t<Sender, receiver_t>;
+
+    // this could throw; if it does, there's nothing to clean up
+    auto opToStart = std::make_unique<manual_lifetime<op_t>>();
+
+    // this could throw; if it does, the only clean-up we need is to destroy
+    // and deallocate the manual_lifetime, which is handled by opToStart's
+    // destructor so we're good
+    opToStart->construct_from([&] {
+      return connect(
+          (Sender&&)sender,
+          receiver_t{
+              (Scheduler&&)scheduler,
+              stopSource_.get_token(),
+              opToStart.get(),
+              this});
+    });
+
+    // At this point, the rest of the function is noexcept, but opToStart's
+    // destructor is no longer enough to properly clean up because it won't
+    // invoke destruct().  We need to ensure that we either call destruct()
+    // ourselves or complete the operation so *it* can call destruct().
+
+    if (try_record_start()) {
+      // start is noexcept so we can assume that the operation will complete
+      // after this, which means we can rely on its self-ownership to ensure
+      // that it is eventually deleted
+      unifex::start(opToStart.release()->get());
+    }
+    else {
+      // we've been stopped so clean up and bail out
+      opToStart->destruct();
+    }
+  }
+
+  [[nodiscard]] auto cleanup() noexcept {
+    return sequence(transform(just(), [this]() noexcept {
+      request_stop();
+    }), evt_.async_wait());
+  }
```

Comment:
The `opState_` field is never loaded with `acquire` semantics which means that releases writes aren't synchronizing with any operation (because only acquire or stronger loads will sync with a release). 

I think that you may want to have the call to cleanup do a final acquire load on the `opState_` to ensure that all the work completed by submitted senders is visible to the thread that calls cleanup. Otherwise the only synchronization point that I see is the last sender to complete will set `evt_` which (I believe) synchronizes with the `async_wait`. I assume that the caller of cleanup would want to see all side affects of the senders that were submitted to this `async_scope`. Without the final acquire, then only the side affects of the last sender to complete will be visible to the caller of cleanup.

```suggestion
  [[nodiscard]] auto cleanup() noexcept {
    return sequence(transform(just(), [this]() noexcept {
      request_stop();
    }), transform(evt_.async_wait(), [this]() noexcept {
      opState_.load(std::memory_order_acquire);
    }));
  }
```